client/termsetting/termsetting.ts

```diff
@@ -847,18 +847,46 @@ type DefaultQByTsHandler = {
 	snplst?: SnpsQ
 }
 
-export async function fillTermWrapper(tw: TermWrapper, vocabApi: VocabApi, defaultQByTsHandler?: DefaultQByTsHandler) {
+export async function fillTwLst(twlst: TwLst, vocabApi: VocabApi, defaultQByTsHandler?: DefaultQByTsHandler): void {
+	const dictTerms = await getDictTerms(twlst, vocabApi)
+	const promises: Promise[] = []
+	for (const tw of twlst) {
+		if (!tw.term) tw.term = dictTerms[tw.id]
+		promises.push(fillTermWrapper(tw, vocabApi, defaultQByTsHandler))
+	}
+	await Promise.all(promises)
+}
+
+async function getDictTerms(twlst: TwLst, vocabApi: VocabApi) {
+	const ids: string[] = []
+	for (const tw of twlst) {
+		// non-dictionary tw would always have a tw.term, so only dictionary tw will be tracked here
+		if (!tw.term) {
+			if (tw.id == undefined || tw.id === '') throw 'missing both .id and .term'
+			ids.push(tw.id)
+		}
+	}
+	// ids only have dictionary terms
+	const terms = ids.length ? await vocabApi.getTerms(ids) : {}
+	for (const id of ids) {
+		if (!terms[id]) throw `missing dictionary term for id=${tw.id}`
+	}
+	return terms
+}
+
+export async function fillTermWrapper(
+	tw: TermWrapper,
+	vocabApi: VocabApi,
+	defaultQByTsHandler?: DefaultQByTsHandler
+): TermWrapper {
 	tw.isAtomic = true
 	if (!tw.$id) tw.$id = get$id()
-
 	if (!tw.term) {
-		if (tw.id == undefined || tw.id === '') throw 'missing both .id and .term'
-		// has .id but no .term, must be a dictionary term
-		// as non-dict term must have tw.term{}
-		tw.term = await vocabApi.getterm(tw.id)
+		const terms = await getDictTerms([tw], vocabApi)
+		tw.term = terms[tw.id]
```

Comment:
tw.id might be undefined

Reply:
@airenzp could you help fix these type errors? my code editor does highlight these for me